### ISSUE TITLE
Range partition on double precision column

### ISSUE THREAD

**Issue by autcrock** [moat/range_partitioning#3]
Hi Corey.

Thank you for this extension.

I am a relative beginner with Postgres.  My question is:

> How do I set up a partitioned table on a double precision column?

Working on Postgres 9.5, I believe that I need to first create a range type:

```
CREATE TYPE doublerange AS RANGE (
    subtype = float8,
    subtype_diff = float8mi
);

```

But I guess there is a step before the following lines:

```
create table part_test.foo_dp( x double precision );
select range_partitioning.create_parent('part_test.foo_dp', 'x', 'range_partitioning');
```

because they fail:

```
ERROR:  query returned more than one row
CONTEXT:  PL/pgSQL function create_parent(text,text,text) line 6 at SQL statement
********** Error **********

ERROR: query returned more than one row
SQL state: P0003
Context: PL/pgSQL function create_parent(text,text,text) line 6 at SQL statement

```

Any help appreciated...

Mike.

**Comment by chuinker** [comment]
That error happens when more than one range type fits the partitioning column
I knew this issue would come up eventually, and I'll have to add an option to specify the desired range type.

**Comment by autcrock** [comment]
That was a fast reply - thanks.

**Comment by chuinker** [comment]
I'll try to get to this issue in the next few days. Time is tight, but I ran into something similar when doing text range types of differing collation sequences, so it's probably to do with a double being able to use a numeric range type in addition to the range type you created...or something like that.

**Comment by coreyatmoat** [comment]
Hmm...I'm not able to re-create your issue.

You do need to create a doublerange type or something equivalent, but it works for me

```shell
$ createdb double_test
$ psql double_test
```
```sql
psql (9.5.0)
Type "help" for help.

# create schema range_partitioning;
CREATE SCHEMA
Time: 2.122 ms
# create extension range_partitioning schema range_partitioning;
CREATE EXTENSION
Time: 41.063 ms
# create table foo( x double precision);
CREATE TABLE
Time: 1.932 ms
# select range_partitioning.create_parent('public.foo','x','public');
ERROR:  P0002: query returned no rows
CONTEXT:  PL/pgSQL function create_parent(text,text,text) line 6 at SQL statement
LOCATION:  exec_stmt_execsql, pl_exec.c:3670
Time: 2.358 ms
# CREATE TYPE doublerange AS RANGE (
... #     subtype = float8,
... #     subtype_diff = float8mi
... # );
CREATE TYPE
Time: 1.936 ms
# select range_partitioning.create_parent('public.foo','x','public');
 create_parent
---------------

(1 row)

# \d public.foo_p0
         Table "public.foo_p0"
 Column |       Type       | Modifiers
--------+------------------+-----------
 x      | double precision |
Check constraints:
    "foo_p0" CHECK (true)
Inherits: foo

# select range_partitioning.create_partition('public.foo','(,0.0]');
 create_partition
------------------

(1 row)

Time: 11.013 ms
# select range_partitioning.create_partition('public.foo','(0.0,1.0)');
 create_partition
------------------

(1 row)

Time: 8.866 ms
# \d+ public.foo
                              Table "public.foo"
 Column |       Type       | Modifiers | Storage | Stats target | Description
--------+------------------+-----------+---------+--------------+-------------
 x      | double precision |           | plain   |              |
Triggers:
    foo_ins_trig BEFORE INSERT ON foo FOR EACH ROW EXECUTE PROCEDURE foo_ins_trigger()
Child tables: foo_p0,
              foo_p1,
              foo_p2

# \d+ public.foo_p0
                            Table "public.foo_p0"
 Column |       Type       | Modifiers | Storage | Stats target | Description
--------+------------------+-----------+---------+--------------+-------------
 x      | double precision |           | plain   |              |
Check constraints:
    "foo_p0" CHECK (x >= '1'::double precision)
Inherits: foo

# \d+ public.foo_p1
                            Table "public.foo_p1"
 Column |       Type       | Modifiers | Storage | Stats target | Description
--------+------------------+-----------+---------+--------------+-------------
 x      | double precision |           | plain   |              |
Check constraints:
    "foo_p1" CHECK (x <= '0'::double precision)
Inherits: foo

# \d+ public.foo_p2
                            Table "public.foo_p2"
 Column |       Type       | Modifiers | Storage | Stats target | Description
--------+------------------+-----------+---------+--------------+-------------
 x      | double precision |           | plain   |              |
Check constraints:
    "foo_p2" CHECK (x > '0'::double precision AND x < '1'::double precision)
Inherits: foo
```

Can you retry these steps on a fresh database and let me know if and where it goes bad for you?

**Comment by autcrock** [comment]
Thanks Corey.

**Comment by coreyatmoat** [comment]
Same Corey as before, btw.

**Comment by autcrock** [comment]
Just logged in from my mobile and saw your SQL above.  I'll be away from computers till Monday and will try It again then.  I'm running the extension in a docker instance so it may be that the installation or some earlier thing I had done had some effect.  The instance is based off the latest 9.5 docker image.  Mike.

**Comment by autcrock** [comment]
OK.  The above script on a clean database runs as yours above.  All as it should.

I remember now that last Friday I had first created a range type called float8range, while learning about range types which, when present in the above series of test statements, causes the error I reported.

Apologies for the false alarm Corey, and thank you for your help.

You are right, though, about adding an additional range type argument and perhaps it would also help if the create_parent error message returned the rows for diagnostic purposes.

**Comment by chuinker** [comment]
That or I need to improve the error message to indicate that there is a conflict, possibly which ranges are in conflict.

**Comment by autcrock** [comment]
Even better!  Feel free to use me as a tester - this is a good extension.

**Comment by coreyatmoat** [comment]
You should be able to test now by pulling from git and upgrading to 1.2.2.

**Comment by autcrock** [comment]
Thanks Corey for your quick response and good work.

The old create_parent syntax works as expected both with one, a duplicate eligible range type and also with none.  In each of the three cases an informative message is printed to the console.

The new syntax works with no eligible types, applies the range type argument correctly and also fails gracefully when directed to implement a partition on a range type which has already been used to partition the same table.

The only glitchy error message I got was when I forgot to put quotes around the range type argument in the new create_parent syntax - perhaps an argument type check would help here:

```
ERROR:  missing FROM-clause entry for table "public"
LINE 20: ...oning.create_parent('public.foo_new','x','public',public.flo...
                                                              ^

********** Error **********

ERROR: missing FROM-clause entry for table "public"
SQL state: 42P01
Character: 723

```

Separately as a data point for you, I used your extension (the old version not the new one) to successfully create 14 partitions on a pre-populated 26 million row geospatial table on a double precision geometry area column.  It took roughly 3.5 hours last night on an AWS m4.4xlarge EC2 instance with Postgres 0.5 running in docker on Ubuntu 14.04.

So thanks again!

Mike.
